src/SourceLocator/Type/MemoizingSourceLocator.php

```diff
@@ -37,17 +38,23 @@ public function locateIdentifier(Reflector $reflector, Identifier $identifier):
             = $this->wrappedSourceLocator->locateIdentifier($reflector, $identifier);
     }
 
-    /** @return list<Reflection> */
-    public function locateIdentifiersByType(Reflector $reflector, IdentifierType $identifierType): array
+    /** @return Generator<Reflection> */
+    public function locateIdentifiersByType(Reflector $reflector, IdentifierType $identifierType): Generator
     {
         $cacheKey = sprintf('%s_%s', $this->reflectorCacheKey($reflector), $this->identifierTypeToCacheKey($identifierType));
 
         if (array_key_exists($cacheKey, $this->cacheByIdentifierTypeKeyAndOid)) {
-            return $this->cacheByIdentifierTypeKeyAndOid[$cacheKey];
+            yield from $this->cacheByIdentifierTypeKeyAndOid[$cacheKey];
+
+            return;
+        }
+
+        $items = [];
+        foreach ($this->wrappedSourceLocator->locateIdentifiersByType($reflector, $identifierType) as $item) {
+            yield $items[] = $item;
```

Comment:
The original implementation caches items before returning them: the new implementation caches items after returning them.

I'm unsure about approach here, but it is a subtle change that can lead to uncached results if the consumer of this iterator crashes